### ISSUE TITLE
TS2719/TS2322 false positive on a generic type forwarded through an interface `extends` boundary (tsc clean, tsgo fails)

### ISSUE THREAD

**Issue by alekseigoloviichuk** [microsoft/typescript-go#4581]
## Steps to reproduce

A generic React wrapper whose props interface **`extends`** a library's generic options type, then spreads those props into the library's generic component. `tsgo` reports the forwarded `columns` prop as *"Two different types with this name exist, but they are unrelated"* — even though **both sides of the diagnostic are the identical type from the identical declaration file**. `tsc` (the JS compiler) accepts it. This reproduces with `material-react-table@2.0.4`; the minimal trigger is the `interface … extends MRT_TableOptions<TData>` combined with the `{...props}` spread.

<details><summary><code>package.json</code></summary>

```json
{
  "name": "tsgo-mrt-false-positive",
  "private": true,
  "dependencies": {
    "@emotion/react": "11.11.3",
    "@emotion/styled": "11.11.0",
    "@mui/icons-material": "5.15.0",
    "@mui/material": "5.15.0",
    "material-react-table": "2.0.4",
    "react": "18.2.0",
    "react-dom": "18.2.0"
  },
  "devDependencies": {
    "@types/react": "18.2.0",
    "typescript": "7.0.2"
  }
}
```
</details>

`tsconfig.json`:
```json
{
  "compilerOptions": {
    "strict": true,
    "target": "ES2021",
    "lib": ["ES2021", "DOM"],
    "module": "ESNext",
    "moduleResolution": "bundler",
    "jsx": "react-jsx",
    "esModuleInterop": true,
    "skipLibCheck": true,
    "noEmit": true
  },
  "include": ["src"]
}
```

`src/Table.tsx`:
```tsx
import { MaterialReactTable } from 'material-react-table';
import type { MRT_TableOptions } from 'material-react-table';

interface TableProps<TData extends Record<string, any> = {}> extends MRT_TableOptions<TData> {
  isLoading?: boolean;
}

export const Table = <TData extends Record<string, any> = Record<string, any>>({
  columns,
  data,
  ...props
}: TableProps<TData>) => {
  return <MaterialReactTable columns={columns} data={data} {...props} />;
};
```

```bash
npm install --legacy-peer-deps
npx tsc --noEmit -p tsconfig.json
```

## Behavior with `typescript@6.0.3` (and 5.x)

Compiles cleanly — exit 0, no diagnostics.

## Behavior with `tsgo`

Tested `typescript@7.0.2`, `typescript@7.1.0-dev.20260708.3`, **and** `typescript@7.1.0-dev.20260710.1` (latest at time of writing) — all produce the same single error:

```
src/Table.tsx(13,30): error TS2719: Type 'MRT_ColumnDef<TData, any>[]' is not assignable to type 'MRT_ColumnDef<TData, any>[]'. Two different types with this name exist, but they are unrelated.
```

Both types are written identically and resolve to the same `material-react-table/dist/index` declaration.

## Notes

- Reproduces with a **single** copy of `@tanstack/react-table` in the tree — it is not a duplicate-package issue.
- The trigger is the combination of (a) `interface TableProps<TData> extends MRT_TableOptions<TData>` and (b) forwarding those props into the generic `<MaterialReactTable>` via `{...props}`. Using `MRT_TableOptions<TData>` directly (without the `extends` indirection), or dropping the spread, makes it disappear.
- Passing more of the generic props explicitly (e.g. `state`, `mui*Props` callbacks that reference `MRT_Row<TData>` / `MRT_TableInstance<TData>`) just multiplies the same diagnostic across those props.
- Likely the same class as #3488 (identical-type `TS2719` under tsgo), but a different trigger — there it was a repeated conditional return type; here it is the identity of a generic instantiation of an imported type resurfaced through `interface extends`.
- Workaround used downstream: cast the callee (`MaterialReactTable as unknown as ComponentType<MRT_TableOptions<TData>>`), which keeps the forwarded props type-checked.

**Comment by alscotty-fox** [comment]
+1 on this issue

**Comment by RyanCavanaugh** [comment]
Minimal repro
```ts
// Table.tsx
import { MaterialReactTable, type MRT_TableOptions } from "./mrt";

interface TableProps<T> extends MRT_TableOptions<T> {}

export const Table = <T,>({ defaultColumn, ...props }: TableProps<T>) =>
  MaterialReactTable({ defaultColumn, ...props });

// mrt.d.ts
type Key<T, Done = false> = Done extends true ? never : unknown extends T ? string : T extends (T extends {} ? T : never) ? never : string;

export type MRT_TableOptions<T> = { defaultColumn: (value: T) => void } & {
  defaultColumn: { Cell: <Done = false>(options: MRT_TableOptions<T>) => Key<T, Done> };
};

export declare function MaterialReactTable<T>(props: MRT_TableOptions<T>): void;
```
Produces
```
Table.tsx(6,24): error TS2719: Type '((value: T) => void) & { Cell: <Done = false>(options: import("Q:/Throwaway/4581/repro-4581/src/mrt").MRT_TableOptions<T>) => import("Q:/Throwaway/4581/repro-4581/src/mrt").Key<T, Done>; }' is not assignable to type '((value: T) => void) & { Cell: <Done = false>(options: import("Q:/Throwaway/4581/repro-4581/src/mrt").MRT_TableOptions<T>) => import("Q:/Throwaway/4581/repro-4581/src/mrt").Key<T, Done>; }'. Two different types with this name exist, but they are unrelated.
  Type '((value: T) => void) & { Cell: <Done = false>(options: MRT_TableOptions<T>) => Key<T, Done>; }' is not assignable to type '{ Cell: <Done = false>(options: MRT_TableOptions<T>) => Key<T, Done>; }'.
    Types of property 'Cell' are incompatible.
      Type '<Done = false>(options: import("Q:/Throwaway/4581/repro-4581/src/mrt").MRT_TableOptions<T>) => import("Q:/Throwaway/4581/repro-4581/src/mrt").Key<T, Done>' is not assignable to type '<Done = false>(options: import("Q:/Throwaway/4581/repro-4581/src/mrt").MRT_TableOptions<T>) => import("Q:/Throwaway/4581/repro-4581/src/mrt").Key<T, Done>'. Two different types with this name exist, but they are unrelated.
        Types of parameters 'options' and 'options' are incompatible.
          Type 'import("Q:/Throwaway/4581/repro-4581/src/mrt").MRT_TableOptions<T>' is not assignable to type 'import("Q:/Throwaway/4581/repro-4581/src/mrt").MRT_TableOptions<T>'. Two different types with this name exist, but they are unrelated.
            Type 'MRT_TableOptions<T>' is not assignable to type '{ defaultColumn: { Cell: <Done = false>(options: MRT_TableOptions<T>) => Key<T, Done>; }; }'.
              The types of 'defaultColumn.Cell(...)' are incompatible between these types.
                Type 'unknown extends T ? string : T extends (T extends {} ? T : never) ? never : string' is not assignable to type 'unknown extends T ? string : T extends (T extends {} ? T : never) ? never : string'. Two different types with this name exist, but they are unrelated.
                  Type 'string | (T extends (T extends {} ? T : never) ? never : string)' is not assignable to type 'unknown extends T ? string : T extends (T extends {} ? T : never) ? never : string'.
                    Type 'string' is not assignable to type 'unknown extends T ? string : T extends (T extends {} ? T : never) ? never : string'.
```

**Comment by RyanCavanaugh** [comment]
Possibly minimal repro (?!), this one at least demonstrates the difference between 6 and 7:
```ts
type IsTuple<TMaybeTuple> = TMaybeTuple extends readonly any[] ? TMaybeTuple : never;

type DeepKeys<TObj, TDepth extends any[] = []> = TDepth extends [any]
  ? never
  : TObj extends readonly any[] & IsTuple<TObj>
    ? never
    : TObj extends object
      ? (keyof TObj & string) | DeepKeysPrefix<TObj, keyof TObj, TDepth>
      : never;

type DeepKeysPrefix<TParent, TPrefix, TDepth extends any[]> = TPrefix extends keyof TParent
  ? DeepKeys<TParent[TPrefix], [...TDepth, any]>
  : never;

type ColumnDef<TCol extends object> = Pick<TCol, never> & {
  accessorKey: DeepKeys<TCol>;
};

type TableOptions<TOpts extends object> = {
  columns: ColumnDef<TOpts>;
  onRowCancel?: (values: Record<DeepKeys<TOpts>, any>) => void;
};

declare function DataTable<TArg extends object>(props: TableOptions<TArg>): void;

export const Table = <TRow extends object>({ columns, ...props }: TableOptions<TRow>) => {
  //                  tsgo reports TS2719 on this argument vvvvvvv
  return DataTable({ columns, ...props });
};
```

```
src/repro.ts:28:22 - error TS2719: Type 'ColumnDef<TRow>' is not assignable to type 'ColumnDef<TRow>'. Two different types with this name exist, but they are unrelated.
  Type 'ColumnDef<TRow>' is not assignable to type '{ accessorKey: TRow extends readonly any[] & IsTuple<TRow> ? never : TRow extends object ? DeepKeysPrefix<TRow, keyof TRow, []> | (keyof TRow & string) : never; }'.
    Types of property 'accessorKey' are incompatible.
      Type 'TRow extends readonly any[] & IsTuple<TRow> ? never : TRow extends object ? DeepKeysPrefix<TRow, keyof TRow, []> | (keyof TRow & string) : never' is not assignable to type 'TRow extends readonly any[] & IsTuple<TRow> ? never : TRow extends object ? DeepKeysPrefix<TRow, keyof TRow, []> | (keyof TRow & string) : never'. Two different types with this name exist, but they are unrelated.
        Type 'TRow extends object ? DeepKeysPrefix<TRow, keyof TRow, []> | (keyof TRow & string) : never' is not assignable to type 'TRow extends readonly any[] & IsTuple<TRow> ? never : TRow extends object ? DeepKeysPrefix<TRow, keyof TRow, []> | (keyof TRow & string) : never'.
          Type 'DeepKeysPrefix<TRow, keyof TRow, []> | (keyof TRow & string)' is not assignable to type 'TRow extends readonly any[] & IsTuple<TRow> ? never : TRow extends object ? DeepKeysPrefix<TRow, keyof TRow, []> | (keyof TRow & string) : never'.
            Type 'keyof TRow & string' is not assignable to type 'TRow extends readonly any[] & IsTuple<TRow> ? never : TRow extends object ? DeepKeysPrefix<TRow, keyof TRow, []> | (keyof TRow & string) : never'.

28   return DataTable({ columns, ...props });
                        ~~~~~~~

  src/repro.ts:20:3 - The expected type comes from property 'columns' which is declared here on type 'TableOptions<TRow>'
    20   columns: ColumnDef<TOpts>;
         ~~~~~~~
```

**Comment by RyanCavanaugh** [comment]
Smaller-ish, AI comments left in for posterity:
```ts
// Repro for microsoft/typescript-go#4581
//
//   tsc  6.0.3 (and 5.x): clean, exit 0
//   tsgo 7.0.2:           error TS2719 on the DataTable(...) call below:
//     Type 'TRow extends any[] & IsTuple<TRow> ? ... : never' is not assignable to
//     type 'TRow extends any[] & IsTuple<TRow> ? ... : never'.
//     Two different types with this name exist, but they are unrelated.
//
// Both sides of the diagnostic are `DeepKeys<TRow>`: the same deferred conditional
// type, from the same declaration, with the same type argument. `TArg` is inferred
// as `TRow`, so the check is `DeepKeys<TRow>` against `DeepKeys<TRow>`; it should
// succeed by identity. tsgo instead structurally distributes the conditional and
// then fails to relate it back to itself.
//
// `DeepKeys` is a cut-down version of @tanstack/table-core's `DeepKeys`, which is
// how the original report hit this through `material-react-table`.
//
// Every piece below is load-bearing -- removing any one of them makes the error
// disappear under tsgo:
//   * the `IsTuple` alias, and the `any[] &` intersection in front of it
//   * the `TDone` recursion cut-off (i.e. `DeepKeys` recurring through `DeepKeysPrefix`)
//   * both arms of the `keyof TObj | DeepKeysPrefix<...>` union
//   * `TRow`'s `extends object` constraint
//   * passing the value through the object literal so `TArg` has to be inferred
//
// Each declaration uses a distinctly named type parameter so the diagnostic is
// readable; the bug does not depend on the names.

type IsTuple<TMaybeTuple> = TMaybeTuple extends any[] ? TMaybeTuple : never;

type DeepKeys<TObj, TDone = false> = TDone extends true
  ? never
  : TObj extends any[] & IsTuple<TObj>
    ? never
    : TObj extends object
      ? keyof TObj | DeepKeysPrefix<TObj, keyof TObj>
      : never;

type DeepKeysPrefix<TParent, TPrefix> = TPrefix extends keyof TParent
  ? DeepKeys<TParent[TPrefix], true>
  : never;

declare function DataTable<TArg>(props: { accessorKey: DeepKeys<TArg> }): void;

export const Table = <TRow extends object>(accessorKey: DeepKeys<TRow>) =>
  //          tsgo reports TS2719 here vvvvvvvvvvv
  DataTable({ accessorKey });
```

**Comment by RyanCavanaugh** [comment]
The agents were having a hard time fixing this bug because it's already been fixed in nightly 🫠
